### ISSUE TITLE
Explore using microservices in the back-end

### ISSUE THREAD

**Issue by humphd** [Seneca-CDOT/telescope#1627]
**What would you like to be added**:

I'd like to explore using microservices in our back-end.  I think the most logical places to begin would be:

1. **auth**: break our authentication and authorization routes out into a separate service, which does SAML based SSO (same as now), but gives back a token that the client can use on subsequent requests.  One of the goals of this re-design is to make it possible for a user to authenticate on production and then use that token in a local build.  In other words: use Seneca's SAML SSO for authentication, and our tokens for authorization within the Telescope ecosystem.
2. **images**: we have various needs around images: providing random backgrounds for the hero component, resizing and/or changing formats of images in posts (e.g., a user posts a 2M image in a Medium post, we should down-scale it automatically when we parse the post) and caching.
3. **autodeployment**: this is already a microservice of sorts.  We should use and evolve it as a basis for building others.

There are lots of others I could extract from our current back-end, but this might be enough to get us started.

**Why would you like this to be added**:

Why do this?  Partly I'm interested in the approach and want to explore it.  I know that @raygervais has some real-world experience doing this, as do I with Mozilla.  Also, I'm finding that our current back-end and data layer are really tightly integrated, and making changes is more complicated than I'd like.

Some goals I have for this work:

- try to maintain the benefits of a single repo (one place to file and discuss issues, easy to share code between parts of the app, one source of truth for dependencies, easier to integrate) but explore how to make it really easy to develop, test, run, and deploy small parts of Telescope without having to work with all of it.
- make sure that we don't take on more DevOps work than we can practically manage.  Already we don't have a good solution here, and we need to think long term.
- stay focused on the skills and abilities of our student developers.  The students I get coming into this work have learned Express, React, Mongo, Node, etc. so we should probably stay in that space vs. jumping into new languages or frameworks.  I'd like people to be able to dive in and not alienate people just to use something different.
- experiment with this in a small way vs. redoing the whole app.  We might go that route by the time we're done, or we might stop with 1 or 2 microservices.  We might also decide a monolith is better.
- do this work in parallel with shipping our monolith.  We should be able to develop this work in parallel to our existing code and flip a switch (env probably) and have things run via services vs. the existing back-end.  I don't want to halt progress to make this happen.

**Starting Points**

Let's begin with some discussion here, on a Teams call, in Slack, etc. and collect some ideas about how we want to approach this.  To kick things off, when I say microservice, I mean this:

- service based approach: think "doing one or a few related things" and "providing a service" vs. "building the whole app"
- loosely coupled, HTTP as the interface between services (i.e., we're building little web servers)
- easy to test on their own: the services route(s) should work in this way (inputs, outputs)
- easy to deploy on their own
- easy to understand in isolation

Moving to this model for some parts of our app means that we can rethink how deployment works.  Shipping "1.6" means something different when you are pushing out different parts of the app at different times.  Integration and deployment become something we have to consider more seriously.

Next, we need some volunteers to get involved with the three microservice migrations above.  Who on the team has interest in this space?

**Comment by raygervais** [comment]
I'd be absolutely interested in helping with this, either in the orchestration or porting area of a service!

**Comment by raygervais** [comment]
For the image MS, you're thinking of handling *all* images through it @humphd? Not just backgrounds / telescope assets? Asking because that may be a data-coupling we'd need to figure out. 

Also asking as I explore creating a POC for said image service as a microservice

**Comment by humphd** [comment]
Here's how I imagine the image service working (details like route names, etc. TBD).  There are at least two things I want:

1. Use it to get optimized, scaled versions of our background images.  I ask for `/background?width=750&type=webp` or something, and I get back one of our Unsplash images at 750px wide as a WebP resource.

2. Use it to downscale and optimize images in posts.  Consider a real example: this recent post by @rjayroso: https://roycedev.medium.com/two-steps-forward-one-step-back-b0c105037a04.  It contains a really nice Unsplash image, https://miro.medium.com/max/4800/1*Inw2S7IHOgI0EfsEs5o0fw.jpeg which is a 4000 x 2666 JPEG weighing 6.09M on the network.  Medium has a service that does almost exactly what I want to build, and it uses https://cdn-images-1.medium.com/max/1024/1*Inw2S7IHOgI0EfsEs5o0fw.jpeg, a 1024 x 682 JPEG weighing 384K.  We could actually compress that down to ~7K as a WebP.  When we parse the RSS feed HTML, we'll have `<img src="original-url>` in posts, and what I'm suggesting is that we could do a rewrite pass of all of these URLs to: `<img src="https://image.telescope.cdot.systems/?url=original-url&width=768&type=webp" data-original-srd="original-src">`.  When that new URL gets requested, one of two things will happen: 1) it will go to the image service and that will trigger a download and resample via sharp or something using streams, and the result will get cached by nginx; 2) nginx will already have the cached version of that optimized image, and will serve it instead.

So the image service doesn't need to retain any data, or know anything about our app.  But our app can be clever about how it uses and reuses these optimized resources.

It's possible that 1) above is a special case of 2), and we just need to hold the URLs to the backgrounds somewhere else, and request them via the image service.  That would mean that some other part of the app deals with picking a random image (URL) to serve from the same endpoint vs. the image service caring about that.

**Comment by humphd** [comment]
But if you want to prototype the image service, we should do that in another issue.  This issue is mostly about exploring the microservices approach with respect to Telescope.

**Comment by humphd** [comment]
Probably a third thing the image service could do: decide if an image at a given URL needs to be optimized or not.  Given a URL to an image, we can get its type, dimensions, etc. and if it crosses some threshold, return an optimized URL, otherwise, return the original URL.  The feed parser could use this to decide when/if to rewrite the image URL(s) for a post.  If it's already optimized, we don't need to bother.

Once upon a time the /planet used to get spammed with massive images from students taking screenshots or pics from their cameras.  I think a lot of modern blogging platforms have mostly solved this for authors, but for cases where it hasn't, we could step in.

**Comment by tonyvugithub** [comment]
If we still need people here, Can I be part of it too? My microservices knowledge is novice at best, so not sure how much useful I could be but definitely would like to learn more and practice it.

**Comment by c3ho** [comment]
Always interested in backend issues!

**Comment by raygervais** [comment]
Awesome, we have volunteers!

@humphd, you've split up three potential candidates for dedicated services:
- Image handling
- Authentication
- Autodeployment

I think the next step is to understand a basic architecture that each micro-service will implement, and then divide the work among volunteers. Thoughts?

By architecture, I'm referring to:
- Dockerfile
- Docker Compose updates
- API Configuration
- Routing Configuration (Gateway Microservice?)

**Comment by manekenpix** [comment]
Count me in!

**Comment by humphd** [comment]
Here's a proposal, let me know what you think:

1. Create a top-level directory for our services, maybe `/api` or `/src/api` if you think we should continue our current layout (I'm happy to get rid of `src` and eventually have `api/` and `client/` etc).
2. Within `/api` we create whatever config files we need to run all the services, maybe that means a docker-compose file, etc.
3. Under `/api` we create directories for each service, for example: `api/image`, `api/deploy`, `api/auth`, etc.  The files related to each service go in those folders, for example: dockerfile, package.json, etc.
4. We add [Traefik](https://traefik.io/) to our production stack, and use it as our api gateway for managing all of our services.  I'm new to Traefik, so there is a learning curve here, but I think it will solve a lot of our problems.  How to integrate it with our current nginx config is another TODO.  When we're doing, you'd be able to access something like `https://api.telescope.cdot.systems` and get at all of the services we have.
5. We should probably do the most basic possible service first.  Maybe the image service, and have it serve a single static image to begin with.  That would let us get all of this infrastructure and templates for building more completed.

Thoughts?

**Comment by humphd** [comment]
First stab at an image service at https://github.com/Seneca-CDOT/telescope/pull/1636.  I'm hoping to use that as a string point to get us going.

**Comment by humphd** [comment]
Filed #1645 to add a Feed Auto Discovery Service, which could be done as another microservice

**Comment by humphd** [comment]
Filed #1646 to add Traefik as our api gateway.

**Comment by raygervais** [comment]
> Here's a proposal, let me know what you think:
> 
>     1. Create a top-level directory for our services, maybe `/api` or `/src/api` if you think we should continue our current layout (I'm happy to get rid of `src` and eventually have `api/` and `client/` etc).
> 
>     2. Within `/api` we create whatever config files we need to run all the services, maybe that means a docker-compose file, etc.
> 
>     3. Under `/api` we create directories for each service, for example: `api/image`, `api/deploy`, `api/auth`, etc.  The files related to each service go in those folders, for example: dockerfile, package.json, etc.
> 
>     4. We add [Traefik](https://traefik.io/) to our production stack, and use it as our api gateway for managing all of our services.  I'm new to Traefik, so there is a learning curve here, but I think it will solve a lot of our problems.  How to integrate it with our current nginx config is another TODO.  When we're doing, you'd be able to access something like `https://api.telescope.cdot.systems` and get at all of the services we have.
> 
>     5. We should probably do the most basic possible service first.  Maybe the image service, and have it serve a single static image to begin with.  That would let us get all of this infrastructure and templates for building more completed.
> 
> 
> Thoughts?

1. I do like the idea housing the services under `/src/`, perhaps a sub folder as you suggest to reduce confusion. 
2. Docker-compose (alongside Traefik) will be the orchestration tool, this means each service will require it's own Dockerfile and have the following requirements: 
   a. Integrated into the API docker-compose file
   b. Leverage documented ports so we aren't using the same port twice
3. Agreed 100%.
4. This will be fun, is there any security concerns with `api.telescope.cdot.systems`? Asking because unless we were to add any form of auth, this would be open to the world.
5. Agreed, reviewing your image POC.

**Comment by humphd** [comment]
Regarding `api.telescope.cdot.systems`, we have a few options:

1. don't put it on the internet, have it be an internal thing only.
2. put it on the internet, since `telescope.cdot.systems/*` already does this and it's been OK.  We have auth for protected routes.

I don't care what we call it, or even if it has a domain of its own.  It might make deployment easier if it's all on the telescope domain.  Not sure.  @manekenpix?

**Comment by raygervais** [comment]
Makes sense to me, @c3ho  @manekenpix  @tonyvugithub I've done a small update to Dave's POC which adds traefik integration along with the semblance of micro-service orchestration here: https://github.com/Seneca-CDOT/telescope/pull/1636

My next step is to get CI happy with the above PR, I figure let's divide up the initial work and tackle them as PRs (if you agree @humphd) against Dave's current PR (which perhaps we'll rename to Microservice POC?).

We need brave developers to tackle: 
- Auth microservice
- Auto deployment
- Kibana integration?????

I'm open to discussion on Teams today when we are all available

**Comment by HyperTHD** [comment]
I know my experience with microservices is very limited at best, but this looks like the makings of something amazing. Signing up to be the brave test developer for this new integration

**Comment by humphd** [comment]
Another easy one to do is #1645 to add a Feed Auto Discovery Service.

I'll take Auth, and collaborate with anyone who wants to help me.

**Comment by raygervais** [comment]
I'd be happy to work with someone on autodeployment!

**Comment by c3ho** [comment]
@humphd happy to learn work + learn about Auth. I should have a bit more availability now to work on Telescope

**Comment by humphd** [comment]
See #1642 for a `user` service we need to build.

**Comment by humphd** [comment]
OK, first service and associated infrastructure for microservices has landed, see #1636 and the image service.  It provides a good example for creating other services (cc @chrispinkney for the user service).

I'm going to close this and follow-up in other issues we have filed.
